Automattic/wp-calypso
# Stats: Add stats list to redux tree

For #5046 this branch seeks to create a new stats state sub-tree to accommodate the data that is currently being managed by the legacy [`stats-list`](https://github.com/Automattic/wp-calypso/tree/master/client/lib/stats/stats-list).  This branch builds upon some of the concepts from @jblz 's prior PR on this topic  ( #2310 ) but also incorporates approaches taken in the Terms state tree as well.

#### Stats List State Structure
Since the stats list houses a variety of data from various endpionts, I have opted to segment the state tree of `items` by `siteId` and the `statsType` - which is essentially the method name used on wpcom.js to call out to a particular stats endpoint ( i.e. `statsStreak`, `statsCountryViews` [etc](https://github.com/Automattic/wpcom.js/blob/master/lib/runtime/site.get.js#L13-L28) ).

`statsTypes` for a givent `siteId` are then keyed by a stringified-query key that allows for persisting the various payloads we use in stats for various time periods and start/end data combinations.  In a similar fashion, the `requesting` sub-tree keeps track of what API requests are in-flight.

```js
{
    items: {
      `siteId`: {
        `statType`: {
          `queryKey`: {} //API payload data
        }
      }
    },
    requesting: {
      `siteId`: {
        `statType`: {
          `queryKey`: `boolean`
        }
      }
    }
}
```

#### Selectors
In the `stats-list` all API payloads are passed through the [`stats-parser`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/stats/stats-list/stats-parser.js) prior to being stored in `localStorage`.  This allows us to massage the data and format it in such a way to be easily rendered by the stats components.

Some ways to handle this with redux would be to do a similar massage of the data in a component ( which felt _odd_ to me ), or I took the approach here of creating a specific memoized selector like what exists in `stats-list`.  In this branch I implemented `getParsedStreakDataForQuery` that displays this approach.  This feels like a good solution to me due to the ease of testing, and the added value of memoizing the result for speed wins.  But please chime in with any opinions here - or alternate approaches.

#### How To Test
This branch does not modify any existing behavior, it is just setting the stage for further redux greatness in calypso stats.  Please refer to the tests in this branch to see the functionality and ensure the tests pass: `npm run test-client client/state/stats `

#### TODO
After this branch is in place a query component could be created, and my first area to migrate will be the stats post streak component on the insights page.
